plugin/trino-druid/src/test/java/io/trino/plugin/druid/TestDruidTypeMapping.java

```diff
@@ -0,0 +1,217 @@
+                .addRoundTrip("col_2", "string", "text_a", createUnboundedVarcharType(), "CAST('text_a' AS varchar)")
+                .addRoundTrip("col_3", "string", "text_b", createUnboundedVarcharType(), "CAST('text_b' AS varchar)")
+                .execute(getQueryRunner(), druidCreateAndInsert("test_unbounded_varchar"));
+    }
+
+    @Test(dataProvider = "timestampValuesProvider")
+    public void testTimestamp(String inputLiteral, String expectedLiteral)
+    {
+        SqlDataTypeTest.create()
+                .addRoundTrip("__time", "timestamp", inputLiteral, TIMESTAMP_MILLIS, expectedLiteral)
+                .addRoundTrip("col_0", "long", "0", BIGINT, "BIGINT '0'")
+                .execute(getQueryRunner(), druidCreateAndInsert("test_timestamp"));
+    }
+
+    @DataProvider
+    public Object[][] timestampValuesProvider()
+    {
+        return new Object[][] {
+                //before epoch
+                {"1958-01-01 13:18:03.123", "TIMESTAMP '1958-01-01 13:18:03.123'"},
+                // after epoch
+                {"2019-03-18 10:01:17.987", "TIMESTAMP '2019-03-18 10:01:17.987'"},
+                // time doubled in JVM zone
+                {"2018-10-28 01:33:17.456", "TIMESTAMP '2018-10-28 01:33:17.456'"},
+                // time doubled in JVM zone
+                {"2018-10-28 03:33:33.333", "TIMESTAMP '2018-10-28 03:33:33.333'"},
+                // epoch
+                {"1970-01-01 00:00:00.000", "TIMESTAMP '1970-01-01 00:00:00.000'"},
+                // time gap in JVM zone
+                {"1970-01-01 00:13:42.000", "TIMESTAMP '1970-01-01 00:13:42.000'"},
+                {"2018-04-01 02:13:55.123", "TIMESTAMP '2018-04-01 02:13:55.123'"},
+                // time gap in Vilnius
+                {"2018-03-25 03:17:17.000", "TIMESTAMP '2018-03-25 03:17:17.000'"},
+                // time gap in Kathmandu
+                {"1986-01-01 00:13:07.000", "TIMESTAMP '1986-01-01 00:13:07.000'"},
+                // test arbitrary time for all supported precisions
+                {"1970-01-01 00:00:00", "TIMESTAMP '1970-01-01 00:00:00.000'"},
+                {"1970-01-01 00:00:00.1", "TIMESTAMP '1970-01-01 00:00:00.100'"},
+                {"1970-01-01 00:00:00.12", "TIMESTAMP '1970-01-01 00:00:00.120'"},
+                {"1970-01-01 00:00:00.123", "TIMESTAMP '1970-01-01 00:00:00.123'"},
+                {"1970-01-01 00:00:00.1239", "TIMESTAMP '1970-01-01 00:00:00.123'"},
+                {"1970-01-01 00:00:00.12399", "TIMESTAMP '1970-01-01 00:00:00.123'"},
+                {"1970-01-01 00:00:00.123999", "TIMESTAMP '1970-01-01 00:00:00.123'"},
+                {"1970-01-01 00:00:00.1239999", "TIMESTAMP '1970-01-01 00:00:00.123'"},
+                {"1970-01-01 00:00:00.12399999", "TIMESTAMP '1970-01-01 00:00:00.123'"},
+                {"1970-01-01 00:00:00.123999999", "TIMESTAMP '1970-01-01 00:00:00.123'"},
```

Comment:
we're getting truncated values? Doesn't seem to match other connectors. Does it also affect predicates?

e.g. `_time <= TIMESTAMP '1970-01-01 00:00:00.123` should not return values like `TIMESTAMP '1970-01-01 00:00:00.123999'`.

Reply:
The time is truncated when it is being ingested by Druid. Even in case of predicate it is being truncated till milliseconds, if it passed as prepared statement or in case of cast it is truncated to seconds precision.

Reply:
Sorry my example was bad. If both the engine and connector are doing truncating then there's no issues.

e.g. `_time <= TIMESTAMP '1970-01-01 00:00:00.1239'` should not return `TIMESTAMP '1970-01-01 00:00:00.124'` (which it could do if we performed rounding - that's why the older `withRounding` mapping function disabled pushdown).

So existing impl is correct - the explanation above is mostly for my own reference. 😄